server/src/main/java/org/elasticsearch/index/mapper/DataStreamTimestampFieldMapper.java

```diff
@@ -0,0 +1,189 @@
+        @Override
+        public Query existsQuery(QueryShardContext context) {
+            throw new IllegalArgumentException("Field [" + name() + "] of type [" + typeName() + "] does not support exists queries");
+        }
+
+    }
+
+    public static class Builder extends MetadataFieldMapper.Builder<Builder> {
+
+        private String fieldName;
+
+        public Builder() {
+            super(NAME, Defaults.DATA_STREAM_TIMESTAMP_FIELD_TYPE);
+        }
+
+        public void setFieldName(String fieldName) {
+            this.fieldName = fieldName;
+        }
+
+        @Override
+        public MetadataFieldMapper build(BuilderContext context) {
+            return new DataStreamTimestampFieldMapper(
+                fieldType,
+                new DataStreamTimestampFieldFieldType(),
+                fieldName
+            );
+        }
+    }
+
+    public static class TypeParser implements MetadataFieldMapper.TypeParser {
+
+        @Override
+        public MetadataFieldMapper.Builder<?> parse(String name,
+                                                    Map<String, Object> node,
+                                                    ParserContext parserContext) throws MapperParsingException {
+            Builder builder = new Builder();
+            for (Iterator<Map.Entry<String, Object>> iterator = node.entrySet().iterator(); iterator.hasNext();) {
+                Map.Entry<String, Object> entry = iterator.next();
+                String fieldName = entry.getKey();
+                Object fieldNode = entry.getValue();
+                if (fieldName.equals("field_name")) {
+                    builder.setFieldName((String) fieldNode);
+                    iterator.remove();
+                }
+            }
+            return builder;
+        }
+
+        @Override
+        public MetadataFieldMapper getDefault(ParserContext parserContext) {
+            return new DataStreamTimestampFieldMapper(Defaults.DATA_STREAM_TIMESTAMP_FIELD_TYPE,
+                new DataStreamTimestampFieldFieldType(), null);
+        }
+    }
+
+    private final String fieldName;
+
+    private DataStreamTimestampFieldMapper(FieldType fieldType, MappedFieldType mappedFieldType, String fieldName) {
+        super(fieldType, mappedFieldType);
+        this.fieldName = fieldName;
+    }
+
+    @Override
+    public void preParse(ParseContext context) throws IOException {
+
+    }
+
+    @Override
+    protected void parseCreateField(ParseContext context) throws IOException {
+
+    }
+
+    @Override
+    public void postParse(ParseContext context) throws IOException {
+        if (fieldName == null) {
+            // not configured, so skip the validation
+            return;
+        }
+
+        IndexableField[] fields = context.rootDoc().getFields(fieldName);
+        if (fields.length == 0) {
+            throw new IllegalArgumentException("data stream timestamp field [" + fieldName + "] is missing");
+        }
+
+        long numberOfValues =
```

Comment:
Maybe we could instead assert that the field type has at least one docvalues field? This would be more robust if we decide to stop indexing points (#https://github.com/elastic/elasticsearch/issues/48665).